lib/delayed/worker.rb

```diff
@@ -89,22 +89,17 @@ def on_exit!
     # Exit early if interrupted.
     def work_off(num = 100)
       success = Concurrent::AtomicFixnum.new(0)
-      failure = Concurrent::AtomicFixnum.new(0)
+      total = 0
 
-      num.times do
+      while total < num
         jobs = reserve_jobs
         break if jobs.empty?
 
+        total += jobs.length
         pool = Concurrent::FixedThreadPool.new(jobs.length)
         jobs.each do |job|
           pool.post do
-            run_thread_callbacks(job) do
```

Comment:
The explanation might make more sense if I put it here next to the related code.

Previously, if `run_thread_callbacks` crashed for any reason, neither `success` nor `failure` would be incremented. Furthermore, the job cleanup (down at the end of `run(job)`) would never occur, so attempts would not be incremented and run_at would not be bumped, making the job immediately available for pickup again.

This is what I mean by the "spinloop". A job gets picked up, its thread crashes, no cleanup occurs, it gets immediately picked up again, etc. Pushing the `run_thread_callbacks` call into `run_job` allows us to clean up the job if, e.g., it fails to deserialize, or if one of our callbacks fails to connect to a secondary resource (as was the case in #41).